docs/architecture/adr-024-coin-metadata.md

```diff
@@ -0,0 +1,145 @@
+account-based model where there are two types of primary accounts -- basic accounts and module accounts.
+All account types have a set of balances that are composed of `Coins`. The `x/bank` module keeps
+track of all balances for all accounts and also keeps track of the total supply of balances in an
+application.
+
+With regards to a balance `amount`, the Cosmos SDK assumes a static and fixed unit of denomination,
+regardless of the denomination itself. In other words, clients and apps built atop a Cosmos-SDK-based
+chain may choose to define and use arbitrary units of denomination to provide a richer UX, however, by
+the time a tx or operation reaches the Cosmos SDK state machine, the `amount` is treated as a single
+unit. For example, for the Cosmos Hub (Gaia), clients assume 1 ATOM = 10^6 uatom, and so all txs and
+operations in the Cosmos SDK work off of units of 10^6.
+
+This clearly provides a poor and limited UX especially as interoperability of networks increases and
+as a result the total amount of asset types increases. We propose to have `x/bank` additionally keep
+track of metadata per `denom` in order to help clients, wallet providers, and explorers improve their
+UX and remove the requirement for making any assumptions on the unit of denomination.
+
+## Decision
+
+The `x/bank` module will be updated to store and index metadata by `denom`, specifically the "base" or
+smallest unit -- the unit the Cosmos SDK state-machine works with. The metadata should ideally be
+extensible and serve the needs of many external consumers and even other modules (e.g. `x/ibc` or `x/nft`).
+At a minimum, the metadata type should include all units and a potential list of denomination aliases.
+
+To provide a means of extensibility, metadata may include a non-zero length list of `extensions`,
+which are simply key/value pairs. In addition, metadata may also include a non-zero length list of
+`aliases` with regard to the origin chain (e.g. `["ETH", "ether"]`). Both `extensions` and `aliases`
+may be changed via governance.
+
+As a result, we can define the type as follows:
+
+```protobuf
+message DenomUnit {
+  string denom    = 1;
+  uint32 exponent = 2;  
+  repeated string aliases = 3;
+}
+
+message Extension {
+  key string = 1;
+  value string = 2;
+}
+
+message Metadata {
+  string description = 1;
+  string base = 2;
+  repeated DenomUnit denom_units = 3;
+  repeated string aliases = 4;
+  repeated Extension extensions = 5;
+}
+```
+
+As an example, the ATOM's metadata can be defined as follows:
+
+```json
+{
+  "description": "The native staking token of the Cosmos Hub.",
+  "base": "uatom",
+  "denom_units": [
+    {
+      "denom": "matom",
+      "exponent": 3,
+      "aliases": [
+        "milliatom"
+      ]
+    },
+    {
+      "denom": "atom",
+      "exponent": 6,
+    }
+  ],
+  "aliases": [
+    "microatom"
+  ],
+  "extensions": [
+    {
+      "key": "total_supply",
+      "value": "543578243452349991"
```

Comment:
updating this extension in the case of `total_supply` would be necessary for the mint module right? For more one-off extension values I imagine customization to other modules is necessary. Is there a way that this could be flexible in the case of the mint module? Maybe the mint module checks a coins metadata for `total_supply` (or other attributes?) before/after minting new coins?

Reply:
@alexanderbez looks like @okwme is also confused with the business logic vs metadata for clients

Reply:
No -- the metadata is not involved in any business logic. `x/mint` would not even change. In this case, `x/mint` already calls `k.bankKeeper.MintCoins`, so `x/bank` would simply update this extension.